### PR TITLE
Pull - Compile Loop : some renames...

### PR DESCRIPTION
**Pull request by diesalbla**
- `F` (the target effect of compilation) => `Final`
- `B` (the value type of compiled) => `Blob`
- `f` (the fold function) => `foldChunk`
- `R` => `Rap` (see the comment)

### CODE REVIEW

File: core/shared/src/main/scala/fs2/Pull.scala [hunk 1]
```diff
@@ -596,133 +596,137 @@ object Pull extends PullLowPriority {
...
+                  out.copy(tail = Translate[src, g, Elem](out.tail, tst.fk))
                 case dd @ Done(_)              => dd
                 case inter @ Interrupted(_, _) => inter
               }
 
-            case uU: Step[f, y] =>
-              val u: Step[G, y] = uU.asInstanceOf[Step[G, y]]
-              val stepScopeF: F[CompileScope[F]] = u.scope match {
-                case None          => F.pure(scope)
+            case uU: Step[f, z] =>
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/Pull.scala:692]: Should `f` be `carrier` here? Since there's no relation to `Final`?

**Comment by mpilquist** [comment]
Mixed feelings on this one:
- :+1: on `Final` and `foldChunk`
- :-1: on `Blob` and `Rap`
- neutral on `K` and `carrier`

In general, I'd prefer to err on the side of single letter names unless the longer name provides some semantic clue to the reader. `Blob` and `Rap` don't provide me any additional context but the fact that they have names makes me think I'm missing something.

**Comment by diesalbla** [comment]
> Can we move go and the definition of R and subtypes somewhere external from the body of compile? Any clarity benefit of doing so?

For the `R` ADT, we would then need to add an extra parameter (because the `CompileScope` is parameterised on final, not carrier. For the compile loop, we would need to add a few type and value parameters from main definition, but can be done.

> Instead of Rap, how about Unroll (and rename go to unroll)? I want to say Step (taken) or StepResult (confusable with Step).

Yeah, `Step` and `StepLeg` are already causing confusion... As for `unroll`, we use that word in the comment of `viewL`, which is closer to the idea of manipulating the data structure. What `go` does is folding and running the Pull structure into the `F` effect, until it reaches a point that sends an event that needs to be handled "up the call stack"... 
- Perhaps `Pop`, which indicates both that we are popping an element from the the recursive call stack, and that something popped out.
- Or `Prompt`, since the what the `R` represent is prompting a "controller" with an stopped point.

**Comment by SystemFw** [comment]
Similar thoughts as Michael tbh. For me the main thing with type parameters is being to see which ones are equal to which, and which change, and single letter names do a way better job for that than long meaningful names (to me), especially because the long names end up not being _that_ meaningful anyway. Perhaps we have a top level comment to establish context instead?

**Comment by diesalbla** [comment]
> [...] the main thing with type parameters is being to see which ones are equal to which, and which change, and single letter names do a way better job for that than long meaningful names (to me), especially because the long names end up not being that meaningful anyway. 

Given three or four variables, maybe... after six or seven the confusion raises too much. Within `go`, there are `F`, `O`, `B`, `G`, `X`, `y`, `g`, `z` , `f`, `h` `q`... To add confusion, some are introduced as type parameters of the constructor, and the compiler itself has some problem deciding if they are patterns or they are new declarations or they are constraints...

The criteria followed here is one that also applies to term or variable names: short names for short-scoped variables,  long names for long-scoped variables. The variables `F`, `G`, and `X`, (or `Final`, `Carrier`, and `Elem`) scopes span over 200 lines, so long names are warranted.

> Perhaps we have a top level comment to establish context instead?

That would help, but readable code is a better code comment.

**Comment by mpilquist** [comment]
Can we unify `R` and `Result`? They are almost identical, but `R` carries a `CompileScope` in all 3 constructors whereas `Result` carries a scope id in one constructor only.

**Comment by diesalbla** [comment]
> Can we unify R and Result? They are almost identical, but R carries a CompileScope in all 3 constructors whereas Result carries a scope id in one constructor only.

AAMOF I was trying in a branch to unify both _Interrupted_ classes, since they are identical in their fields and in a few places we just convert one into the other... but I left it after a while. I saw it could be a bit ubly, to have overlapping ADTs, and it has been past practice not to microoptimise unfrequent cases of failure or interruption.

In general, R and Result cannot be unify, because they represent completely different notions. The `Result` is a type of Pull, and therefore Streams, data structure. It can be used externally, pass through, or stored. The `R` classes are an ephemeral data structure, which we used to convey information through phases/stages/levels of the interpretation execution, and whose life-time is bound to that process.

**Comment by SystemFw** [comment]
> One criteria followed here is one that also applies to term or variable names: short names for short-scoped variables, long names for long-scoped variables. The variables F, G, and X, (or Final, Carrier, and Elem) scopes span over 200 lines, so long names are warranted.

do you know what, I actually think that for type params you want the _opposite_ rule, because fundamentally you are doing some different: unification. For long-scoped variables, whose unification you need to keep track of for a long time, short names keep the signatures uncluttered. For short-scoped ones (where you tend to lose track of where you are in the stack of type variables in your head), a name that reminds you what they are can be helpful

**Comment by SystemFw** [comment]
I'd be in favour of renaming `R` though, a single letter for a concrete type (instead of a type param) is confusing. `Rap` doesn't do it for me though :)

**Comment by SystemFw** [comment]
It's also a shame that shadowing will trigger warning, cause I think it can be helpful in things like `innerMapOutput`

**Comment by diesalbla** [comment]
Closed in favour of #2110

**Pull request closed without merging**